File: ScoreFrame/ScoreFrame/Models/Match.swift
import Foundation
import SwiftData

@Model
final class Match {
    var id: UUID
    var homeTeamName: String
    var awayTeamName: String
    var videoBookmark: Data?
    var videoBookmarksData: Data?
    var createdAt: Date
    var scoreboardStyleData: Data?
    var timerStartTime: TimeInterval?   // キックオフの動画内タイムスタンプ（後方互換用）
    var timerStopTime: TimeInterval?    // 試合終了の動画内タイムスタンプ（後方互換用）
    var timerStartOffset: TimeInterval? // タイマー開始時の試合経過時間（後方互換用）
    var timerSegmentsData: Data?        // [TimerSegment] を JSON エンコード保存
    var matchInfo: String?              // 大会名・日程などの試合情報
    var pkKicksData: Data?              // [PKKick] を JSON エンコード保存
    var penaltyTimersData: Data?       // [PenaltyTimer] を JSON エンコード保存
    var timeoutsData: Data?            // [TimeoutEvent] を JSON エンコード保存
    var skipOverlay: Bool = false       // スコアボードオーバーレイを付けず動画のみ結合

    @Relationship(deleteRule: .cascade, inverse: \ScoreEvent.match)
    var scoreEvents: [ScoreEvent]

    var scoreboardStyle: ScoreboardStyle {
        get {
            guard let data = scoreboardStyleData,
                  let style = try? JSONDecoder().decode(ScoreboardStyle.self, from: data) else {
                return ScoreboardStyle()
            }
            return style
        }
        set {
            scoreboardStyleData = try? JSONEncoder().encode(newValue)
        }
    }

    var videoURL: URL? {
        get {
            guard let bookmark = videoBookmark else { return nil }
            var isStale = false
            return try? URL(
                resolvingBookmarkData: bookmark,
                bookmarkDataIsStale: &isStale
            )
        }
        set {
            videoBookmark = try? newValue?.bookmarkData()
        }
    }

    var videoURLs: [URL] {
        get {
            // 新形式: videoBookmarksData から復元
            if let data = videoBookmarksData,
               let bookmarks = try? JSONDecoder().decode([Data].self, from: data) {
                let urls = bookmarks.compactMap { bookmark -> URL? in
                    var isStale = false
                    return try? URL(
                        resolvingBookmarkData: bookmark,
                        bookmarkDataIsStale: &isStale
                    )
                }
                if !urls.isEmpty { return urls }
            }
            // 旧形式フォールバック: 単一の videoBookmark
            if let url = videoURL {
                return [url]
            }
            return []
        }
        set {
            let bookmarks = newValue.compactMap { try? $0.bookmarkData() }
            videoBookmarksData = try? JSONEncoder().encode(bookmarks)
            // 後方互換: 最初のURLを旧プロパティにも保存
            videoBookmark = bookmarks.first
        }
    }

    var timerSegments: [TimerSegment] {
        get {
            if let data = timerSegmentsData,
               let segments = try? JSONDecoder().decode([TimerSegment].self, from: data),
               !segments.isEmpty {
                return segments
            }
            // 後方互換: 既存の単一タイマーからセグメント1つを自動生成
            if timerStartTime != nil || timerStopTime != nil || timerStartOffset != nil {
                return [TimerSegment(
                    periodLabel: scoreboardStyle.periodLabel,
                    timerStartTime: timerStartTime,
                    timerStopTime: timerStopTime,
                    timerStartOffset: timerStartOffset
                )]
            }
            return []
        }
        set {
            timerSegmentsData = try? JSONEncoder().encode(newValue)
        }
    }

    func segmentIndex(at videoTime: TimeInterval) -> Int? {
        for (i, seg) in timerSegments.enumerated() {
            guard let start = seg.effectiveStartTime else { continue }
            let end = seg.timerStopTime ?? .infinity
            if videoTime >= start && videoTime <= end {
                return i
            }
        }
        return nil
    }

    func currentPeriodLabel(at videoTime: TimeInterval) -> String? {
        guard let idx = segmentIndex(at: videoTime) else {
            // セグメント外 → 直前のセグメントのラベルを返す
            var lastLabel: String?
            for seg in timerSegments {
                guard let start = seg.effectiveStartTime else { continue }
                if start <= videoTime {
                    lastLabel = seg.periodLabel
                }
            }
            return lastLabel
        }
        return timerSegments[idx].periodLabel
    }

    var homeScore: Int {
        scoreEvents.filter { $0.team == .home }.count
    }

    var awayScore: Int {
        scoreEvents.filter { $0.team == .away }.count
    }

    var sortedEvents: [ScoreEvent] {
        scoreEvents.sorted { $0.timestamp < $1.timestamp }
    }

    init(homeTeamName: String, awayTeamName: String) {
        self.id = UUID()
        self.homeTeamName = homeTeamName
        self.awayTeamName = awayTeamName
        self.createdAt = Date()
        self.scoreEvents = []
    }

    var pkKicks: [PKKick] {
        get {
            guard let data = pkKicksData,
                  let kicks = try? JSONDecoder().decode([PKKick].self, from: data) else {
                return []
            }
            return kicks
        }
        set {
            pkKicksData = try? JSONEncoder().encode(newValue)
        }
    }

    var homePKKicks: [PKKick] {
        pkKicks.filter { $0.team == .home }.sorted { $0.order < $1.order }
    }

    var awayPKKicks: [PKKick] {
        pkKicks.filter { $0.team == .away }.sorted { $0.order < $1.order }
    }

    var homePKScore: Int {
        pkKicks.filter { $0.team == .home && $0.isGoal }.count
    }

    var awayPKScore: Int {
        pkKicks.filter { $0.team == .away && $0.isGoal }.count
    }

    func pkKicksAt(time: TimeInterval) -> [PKKick] {
        pkKicks.filter { $0.timestamp <= time }
    }

    func activeTimerSegment(at videoTime: TimeInterval) -> TimerSegment? {
        if let idx = segmentIndex(at: videoTime) {
            return timerSegments[idx]
        }
        // セグメント外 → 直前のアクティブセグメントを返す
        var lastSeg: TimerSegment?
        for seg in timerSegments {
            guard let start = seg.effectiveStartTime else { continue }
            if start <= videoTime { lastSeg = seg }
        }
        return lastSeg
    }

    // MARK: - Penalty Timers

    var penaltyTimers: [PenaltyTimer] {
        get {
            guard let data = penaltyTimersData,
                  let timers = try? JSONDecoder().decode([PenaltyTimer].self, from: data) else {
                return []
            }
            return timers
        }
        set {
            penaltyTimersData = try? JSONEncoder().encode(newValue)
        }
    }

    func activePenaltyTimers(at videoTime: TimeInterval, for team: Team) -> [PenaltyTimer] {
        penaltyTimers
            .filter { $0.team == team && $0.remainingSeconds(at: videoTime) != nil }
            .sorted { $0.timestamp < $1.timestamp }
    }

    // MARK: - Timeouts

    var timeouts: [TimeoutEvent] {
        get {
            guard let data = timeoutsData,
                  let events = try? JSONDecoder().decode([TimeoutEvent].self, from: data) else {
                return []
            }
            return events
        }
        set {
            timeoutsData = try? JSONEncoder().encode(newValue)
        }
    }

    func timeoutCount(for team: Team, at videoTime: TimeInterval) -> Int {
        timeouts.filter { $0.team == team && $0.timestamp <= videoTime }.count
    }

    func isTimeoutActive(at videoTime: TimeInterval) -> Bool {
        timeouts.contains { $0.isActive(at: videoTime) }
    }

    func scoreAt(time: TimeInterval) -> (home: Int, away: Int) {
        var home = 0
        var away = 0
        for event in sortedEvents where event.timestamp <= time {
            switch event.team {
            case .home: home += 1
            case .away: away += 1
            }
        }
        return (home, away)
    }
}

File: ScoreFrame-Android/app/src/main/java/com/scoreframe/model/Match.kt
package com.scoreframe.model

import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json

private val json = Json { ignoreUnknownKeys = true }

data class Match(
    val id: String = java.util.UUID.randomUUID().toString(),
    var homeTeamName: String,
    var awayTeamName: String,
    val createdAt: Long = System.currentTimeMillis(),
    var scoreboardStyle: ScoreboardStyle = ScoreboardStyle(),
    var timerSegments: List<TimerSegment> = emptyList(),
    var matchInfo: String? = null,
    var pkKicks: List<PKKick> = emptyList(),
    var penaltyTimers: List<PenaltyTimer> = emptyList(),
    var timeouts: List<TimeoutEvent> = emptyList(),
    var videoUris: List<String> = emptyList(),
    var skipOverlay: Boolean = false,
    var scoreEvents: List<ScoreEvent> = emptyList()
) {
    val homeScore: Int
        get() = scoreEvents.count { it.team == Team.home }

    val awayScore: Int
        get() = scoreEvents.count { it.team == Team.away }

    val sortedEvents: List<ScoreEvent>
        get() = scoreEvents.sortedBy { it.timestamp }

    val homePKKicks: List<PKKick>
        get() = pkKicks.filter { it.team == Team.home }.sortedBy { it.order }

    val awayPKKicks: List<PKKick>
        get() = pkKicks.filter { it.team == Team.away }.sortedBy { it.order }

    val homePKScore: Int
        get() = pkKicks.count { it.team == Team.home && it.isGoal }

    val awayPKScore: Int
        get() = pkKicks.count { it.team == Team.away && it.isGoal }

    fun scoreAt(time: Double): Pair<Int, Int> {
        var home = 0
        var away = 0
        for (event in sortedEvents) {
            if (event.timestamp > time) break
            when (event.team) {
                Team.home -> home++
                Team.away -> away++
            }
        }
        return home to away
    }

    fun pkKicksAt(time: Double): List<PKKick> {
        return pkKicks.filter { it.timestamp <= time }
    }

    fun segmentIndex(videoTime: Double): Int? {
        for ((i, seg) in timerSegments.withIndex()) {
            val start = seg.effectiveStartTime ?: continue
            val end = seg.timerStopTime ?: Double.MAX_VALUE
            if (videoTime in start..end) return i
        }
        return null
    }

    fun currentPeriodLabel(videoTime: Double): String? {
        val idx = segmentIndex(videoTime)
        if (idx != null) return timerSegments[idx].periodLabel
        var lastLabel: String? = null
        for (seg in timerSegments) {
            val start = seg.effectiveStartTime ?: continue
            if (start <= videoTime) lastLabel = seg.periodLabel
        }
        return lastLabel
    }

    fun activeTimerSegment(videoTime: Double): TimerSegment? {
        val idx = segmentIndex(videoTime)
        if (idx != null) return timerSegments[idx]
        var lastSeg: TimerSegment? = null
        for (seg in timerSegments) {
            val start = seg.effectiveStartTime ?: continue
            if (start <= videoTime) lastSeg = seg
        }
        return lastSeg
    }

    fun activePenaltyTimers(videoTime: Double, forTeam: Team): List<PenaltyTimer> {
        return penaltyTimers
            .filter { it.team == forTeam && it.remainingSeconds(videoTime) != null }
            .sortedBy { it.timestamp }
    }

    fun timeoutCount(forTeam: Team, videoTime: Double): Int {
        return timeouts.count { it.team == forTeam && it.timestamp <= videoTime }
    }

    fun isTimeoutActive(videoTime: Double): Boolean {
        return timeouts.any { it.isActive(videoTime) }
    }

    fun toEntity(): MatchEntity = MatchEntity(
        id = id,
        homeTeamName = homeTeamName,
        awayTeamName = awayTeamName,
        createdAt = createdAt,
        scoreboardStyleJson = json.encodeToString(ScoreboardStyle.serializer(), scoreboardStyle),
        timerSegmentsJson = json.encodeToString(ListSerializer(TimerSegment.serializer()), timerSegments),
        matchInfo = matchInfo,
        pkKicksJson = json.encodeToString(ListSerializer(PKKick.serializer()), pkKicks),
        penaltyTimersJson = json.encodeToString(ListSerializer(PenaltyTimer.serializer()), penaltyTimers),
        timeoutsJson = json.encodeToString(ListSerializer(TimeoutEvent.serializer()), timeouts),
        videoUrisJson = json.encodeToString(ListSerializer(String.serializer()), videoUris),
        skipOverlay = skipOverlay
    )

    companion object {
        fun fromEntity(entity: MatchEntity, events: List<ScoreEvent> = emptyList()): Match {
            return Match(
                id = entity.id,
                homeTeamName = entity.homeTeamName,
                awayTeamName = entity.awayTeamName,
                createdAt = entity.createdAt,
                scoreboardStyle = entity.scoreboardStyleJson?.let {
                    try { json.decodeFromString(ScoreboardStyle.serializer(), it) } catch (_: Exception) { null }
                } ?: ScoreboardStyle(),
                timerSegments = entity.timerSegmentsJson?.let {
                    try { json.decodeFromString(ListSerializer(TimerSegment.serializer()), it) } catch (_: Exception) { null }
                } ?: emptyList(),
                matchInfo = entity.matchInfo,
                pkKicks = entity.pkKicksJson?.let {
                    try { json.decodeFromString(ListSerializer(PKKick.serializer()), it) } catch (_: Exception) { null }
                } ?: emptyList(),
                penaltyTimers = entity.penaltyTimersJson?.let {
                    try { json.decodeFromString(ListSerializer(PenaltyTimer.serializer()), it) } catch (_: Exception) { null }
                } ?: emptyList(),
                timeouts = entity.timeoutsJson?.let {
                    try { json.decodeFromString(ListSerializer(TimeoutEvent.serializer()), it) } catch (_: Exception) { null }
                } ?: emptyList(),
                videoUris = entity.videoUrisJson?.let {
                    try { json.decodeFromString(ListSerializer(String.serializer()), it) } catch (_: Exception) { null }
                } ?: emptyList(),
                skipOverlay = entity.skipOverlay,
                scoreEvents = events
            )
        }
    }
}
